variants/github_actions_ci/workflows/ci.yml

```diff
@@ -0,0 +1,157 @@
+  DATABASE_URL: postgres://postgres:postgres@localhost:5432/postgres
+  SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK }}
+  SLACK_CHANNEL: TODO
+  SLACK_ICON: https://github.com/ackama.png?size=48
+  SLACK_USERNAME: TODO_APP_NAME CI
+  SLACK_FOOTER: ""
+jobs:
+  js_based_checks:
+    runs-on: ubuntu-latest
+    timeout-minutes: 15
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v3
+        with:
+          node-version-file: '.node-version'
+          cache: "yarn"
+      - run: yarn install --frozen-lockfile
+      - run: yarn run js-lint
+      - run: yarn run format-check
+  ruby_based_checks:
+    runs-on: ubuntu-latest
+    timeout-minutes: 20
+    services:
+      db:
+        image: postgres:latest
+        ports: ["5432:5432"]
+        env:
+          POSTGRES_USER: postgres
+          POSTGRES_PASSWORD: postgres
+          POSTGRES_DB: postgres
+        options: >-
+          --health-cmd pg_isready
+          --health-interval 10s
+          --health-timeout 5s
+          --health-retries 5
+      # redis:
+      #   image: redis:latest
+      #   ports:
+      #     - 6379:6379
+      # elasticsearch:
+      #   image: elasticsearch:latest
+      #   env:
+      #     cluster.name: whats-on-test-cluster
+      #     discovery.type: single-node
+      #     ES_JAVA_OPTS: "-Xms1g -Xmx1g"
+      #   ports:
+      #     - 9200:9200
+      #     - 9300:9300
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v2
+      - name: Install Required OS Packages
+        run: sudo apt-get -yqq install libpq-dev
+      - name: Set up Ruby (version given by .ruby-version)
+        uses: ruby/setup-ruby@v1
+        with:
+          bundler-cache: true
+      - run: bundle exec rubocop
+      - run: bundle exec brakeman --run-all-checks --exit-on-warn --format plain .
+      - run: bundle exec bundle audit --update
+      - run: bin/setup
+      - run: bundle exec rspec spec --format progress
+  deploy_to_ec2_via_ssm_and_oidc:
+    name: Deploy with Capistrano over AWS SSM, authenticated by Github OIDC
+    runs-on: ubuntu-latest
+    needs:
+      - ruby_based_checks
+      - js_based_checks
+    timeout-minutes: 15
+    # These permissions are needed to interact with GitHub's OIDC Token endpoint.
+    permissions:
+      id-token: write
+      contents: read
+    steps:
+      # Defines the variables needed for future steps, based off of the
+      # github.ref, indicating which environment we are deploying into.
+      - name: Set variables
+        id: variables
+        run: |
+          if [[ ${{ github.ref }} == 'refs/heads/production' ]]; then
+            echo ::set-output name=ENVIRONMENT::production
+            echo ::set-output name=AWS_ROLE_TO_ASSUME::ClientNameProductionGHARole
+          elif [[ ${{ github.ref }} == 'refs/heads/main' ]]; then
+            echo ::set-output name=ENVIRONMENT::staging
+            echo ::set-output name=AWS_ROLE_TO_ASSUME::ClientNameStagingGHARole
```

Comment:
@G-Rath do you prefer this style (setting variables to choose which env to deploy to) or having separate jobs for deploying to each env?

Reply:
I think I prefer the other style because then you only need one condition, whereas this way you have to think about each step in the job; plus then we should be able to do [workflow reuse](https://docs.github.com/en/actions/using-workflows/reusing-workflows).